src/dune_pkg/fetch.ml

```diff
@@ -188,18 +188,78 @@ type failure =
 
 let label = "dune-fetch"
 
+module Single_run_cache = struct
+  type entry =
+    { filename : Filename.t
+    ; download_complete : unit Fiber.t
+    }
+
+  (* A cache of downloaded files keyed by urls which persists within a single
+     invocation of dune and is deleted when dune terminates *)
+  type t =
+    { temp_dir : Path.t
+    ; url_to_filename : entry String.Map.t ref
+    }
+
+  let create () =
+    let temp_dir = Temp.create Dir ~prefix:"dune" ~suffix:"fetch-archive-cache" in
+    at_exit (fun () -> Temp.destroy Dir temp_dir);
+    { temp_dir; url_to_filename = ref String.Map.empty }
+  ;;
+
+  (* [with_ t ~url ~f] checks whether [url] corresponds to an entry in the
+     cache. If an entry is present then the name of the cached file is
+     returned. Otherwise [f] is called on a path and is expected to download
+     the file at [url] and store it at that path. If [f] returns [Ok ()] but it
+     fails to do create a file at the given path then a [Code_error] is raised.
```

Comment:
I wonder, would it not be better if `f` returned either the path to which the file was downloaded (and then the caller moves it to the cache) or the contents? That way instead of `()` the return value can be something meaningful.

Reply:
I think it's simpler to have `f` create the file at a specified path rather than letting `f` choose its own path. This allows the cached files to be stored in a centralized location of the cache's choosing and in some cases clients of the cache will be able to use the cached files without needing to copy them elsewhere (such as when computing their checksums).